File: check-links.js
const fs = require('fs');
const path = require('path');

const outDir = path.join(__dirname, 'out');
const htmlFiles = [];

function walk(dir) {
  try {
    const items = fs.readdirSync(dir, {withFileTypes: true});
    for (const item of items) {
      const full = path.join(dir, item.name);
      if (item.isDirectory()) {
        walk(full);
      } else if (item.name.endsWith('.html')) {
        htmlFiles.push(full);
      }
    }
  } catch (e) {
    console.error('Error reading dir', dir, e.message);
  }
}
walk(outDir);
console.log(`Found ${htmlFiles.length} HTML files`);

const hrefRegex = /href=["']([^"']+)["']/gi;
let broken = 0;
const brokenLinks = [];

for (const file of htmlFiles) {
  try {
    const content = fs.readFileSync(file, 'utf8');
    let match;
    while ((match = hrefRegex.exec(content)) !== null) {
      const link = match[1];
      if (!link || link.startsWith('javascript:') || link.startsWith('mailto:') || link.startsWith('tel:') || link.startsWith('#')) {
        continue;
      }
      try {
        const url = new URL(link);
        if (!url.protocol || !url.host) {
          // relative link
          let rel = url.pathname;
          // Remove query and hash if any (already removed by regex? but safe)
          const queryIndex = rel.indexOf('?');
          if (queryIndex !== -1) {
            rel = rel.substring(0, queryIndex);
          }
          const hashIndex = rel.indexOf('#');
          if (hashIndex !== -1) {
            rel = rel.substring(0, hashIndex);
          }
          if (rel.startsWith('/')) {
            rel = rel.slice(1);
          }
          if (!rel) {
            rel = 'index.html';
          }
          const candidate = path.join(outDir, rel);
          if (!fs.existsSync(candidate)) {
            // If it's a directory, try index.html inside
            if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) {
              const indexPath = path.join(candidate, 'index.html');
              if (!fs.existsSync(indexPath)) {
                brokenLinks.push({file, link, tried: [candidate, indexPath]});
                broken++;
              }
            } else if (!fs.existsSync(candidate)) {
              brokenLinks.push({file, link, tried: candidate});
              broken++;
            }
          }
        }
        // else: absolute link, we skip for now
      } catch (e) {
        // Invalid URL, skip
        continue;
      }
    }
  } catch (e) {
    console.error('Error reading file', file, e.message);
  }
}

console.log(`Total broken internal links: ${broken}`);
if (broken > 0) {
  console.log('First 10 broken links:');
  for (let i = 0; i < Math.min(brokenLinks.length, 10); i++) {
    const {file, link, tried} = brokenLinks[i];
    console.log(`  ${file}: ${link} -> tried ${tried.join(' or ')}`);
  }
}

// Write report to file
const reportPath = path.join(__dirname, 'broken-links-report.json');
fs.writeFileSync(reportPath, JSON.stringify(brokenLinks, null, 2));
console.log(`Report written to ${reportPath}`);

File: check_links.py
import os
import re
from urllib.parse import urlparse

out_dir = os.path.join(os.path.dirname(__file__), 'out')
html_files = []
for root, dirs, files in os.walk(out_dir):
    for file in files:
        if file.endswith('.html'):
            html_files.append(os.path.join(root, file))

print(f"Found {len(html_files)} HTML files")

href_pattern = re.compile(r'href=["\']([^"\']+)["\']', re.IGNORECASE)
broken = 0
broken_links = []

for html_file in html_files:
    try:
        with open(html_file, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Could not read {html_file}: {e}")
        continue
    for match in href_pattern.finditer(content):
        link = match.group(1)
        # Skip empty, javascript:, mailto:, tel:, # anchors
        if not link or link.startswith('javascript:') or link.startswith('mailto:') or link.startswith('tel:') or link.startswith('#'):
            continue
        # Parse to remove query/fragment
        parsed = urlparse(link)
        path = parsed.path
        if not path:
            continue
        # Determine if internal (no scheme)
        if parsed.scheme:
            # external, skip for now
            continue
        # Normalize path: remove leading slash
        if path.startswith('/'):
            path = path[1:]
        # If empty after stripping slash, it's root -> should exist as index.html
        if not path:
            candidate = os.path.join(out_dir, 'index.html')
            if not os.path.exists(candidate):
                broken_links.append((html_file, link, candidate))
                broken += 1
            continue
        # Check if file exists as is
        candidate = os.path.join(out_dir, path)
        if os.path.exists(candidate):
            continue
        # If not, maybe it's a directory without trailing slash; try adding index.html
        candidate2 = os.path.join(out_dir, path, 'index.html')
        if os.path.exists(candidate2):
            continue
        # Still not found -> broken
        broken_links.append((html_file, link, candidate))
        broken += 1

print(f"\nFound {broken} potentially broken internal links:")
# Show first 20
for src, link, cand in broken_links[:20]:
    print(f"  In {os.path.relpath(src, out_dir)}: href=\"{link}\" -> missing {os.path.relpath(cand, out_dir)}")
if len(broken_links) > 20:
    print(f"  ... and {len(broken_links)-20} more")
# Write report
import json
report_path = os.path.join(os.path.dirname(__file__), 'broken-links-report.json')
with open(report_path, 'w') as f:
    json.dump(broken_links, f, indent=2)
print(f"\nReport written to {report_path}")